### ISSUE TITLE
Extremely slow movement and highlighting on long lines

### ISSUE THREAD

**Issue by dumblob** [neovim/neovim#4736]
- Neovim version: 0.1.4
- [ ] Vim behaves differently? Vim version: 7.4.1797-1
- Operating system/version: Arch Linux 4.4.8-1-lts
- Terminal name/version: xterm
- `$TERM`: xterm

### Actual behaviour

Extremely slow cursor movement in files with long lines (e.g. [englishdictionary.sql](https://sourceforge.net/projects/mysqlenglishdictionary/files/englishdictionary.sql/download )). Apparently caused by very inefficient highlighting (slow regexps?).

### Expected behaviour

Cursor movement at normal speed (i.e. without getting stuck indefinitely).

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC englishdictionary.sql
```

Vim behavior is the same, i.e. bad.

The reason why I'm raising this issue is to propose some timeout for highlighting. I'm working quite a lot with such long-line files and it's pain without highlighting and even more with it.

Even some hack as switching off highlighting for lines longer than some configurable *N* would be fine, because my goal is to edit the file and work with it (and with the indefinitely hung nvim it's impossible to run `:syntax off`).

**Comment by dumblob** [comment]
Might be slightly related to https://github.com/neovim/neovim/issues/614 , https://github.com/neovim/neovim/issues/650 , https://github.com/neovim/neovim/issues/2401 .

**Comment by justinmk** [comment]
> propose some timeout for highlighting

Yes that would be useful, especially for scrolling. Bram sketched out a design in a recent [vim_dev thread](https://groups.google.com/d/msg/vim_dev/qtw5LdCrHJQ/aOeysSR0DgAJ):

```
> > > Can we make it work with vim commands? What comes to mind would be to 
> > > have syntax highlighting work asynchronously, so Vim would stay 
> > > responsive for languages that have known to be slow syntax highlighting. 
> > 
> > Hmm, that's something completely different.  A simplistic version would 
> > be to first redraw the screen in B&W, remembering what was put on the 
> > screen, and then loop over those lines to add highlighting, bailing out 
> > when a key is typed.  Might actually work. 
> 
> This hasn't been done yet, correct? 

Nope.  It could be done like we call timers while waiting for a 
character to be typed.  But storing the state of what screen lines still 
need to be colored, and making sure the coloring is interrupted by a 
typed character, is still some work. And when it takes long it also 
requires timers to be invoked... 
```

Neovim may be in a slightly better position to achieve this because libuv callbacks can "interrupt" normally instead of having to add "yields" everywhere.

**Comment by dumblob** [comment]
The sketched idea sounds feasible.

**Comment by rr-** [comment]
`:syntax off` seem to help only a little bit for JSON files. Maybe it's the parenthesis matching?

**Comment by justinmk** [comment]
@rr- That's likely. Would be a good idea to modify matchparen.vim so that it disables itself on long lines.

**Comment by vincentzhezhang** [comment]
It is also painfully slow for some plugins which only shows part of the lines, e.g. vim-fugitive's :Gblame

**Comment by albertnetymk** [comment]
`:syntax off` works great for long lines. It would be ideal if `syntax off` could be applied for those lines only, instead of for the whole buffer.

**Comment by ronisbr** [comment]
I am seeing this bug when `cursoline` is enabled in Julia source-codes. When I disable `cursorline` or syntax highlight, it seems to work OK. Is it the same bug?

**Comment by nyngwang** [comment]
This is because the highlighting of matched parentheses, try:

```
set nosm
let g:loaded_matchparen=1
```

which will speed it up.

**Comment by frixaco** [comment]
Issue still exists, can confirm it. Tried `nvim -u NONE` and profiling with `:profile start/func */file *`. In my case, I have DB SQL backup file and records have **very** long JSON columns which results in **very** long lines. And it seems like the issue is in runtime [matchparen plugin](https://github.com/neovim/neovim/blob/master/runtime/plugin/matchparen.vim).

I literally cannot edit the file and each action takes a few minutes. I'm using latest nightly Neovim. (Unfortunately) VS Code is completely fine and I can jump around/navigate through the file without any issues (even with the Vim extension). Turning `syntax off` solved the issue, but ofc this is not a solution.

Let me know if you need any details, I can provide anything I can (although, it looks like the issue will never be addressed)

**Comment by lewis6991** [comment]
Is this fixed with #30134?

**Comment by justinmk** [comment]
There are expensive, internal things done unnecessarily on each cursor moved, not only matchparen. There is still lots of room for improvement here.

The general idea of skipping highlight updates entirely until cursor is idle for xxx milliseconds, is very high-leverage and we absolutely should do that.

**Comment by lewis6991** [comment]
It would be good if these things could be quantified.

**Comment by nick4tech** [comment]
@zeertzjq @justinmk @llllvvuu Forgot to tag you on my previous comment ☝

**Comment by clason** [comment]
Vim's whole internal buffer handling ("ropes at home") is not really optimized for long lines, so this is not a single bug to be fixed. Current effort is concentrated around https://github.com/neovim/neovim/pull/30869 and https://github.com/neovim/neovim/pull/22420.

**Comment by clason** [comment]
This is a very old issue that is too broad to be simply actionable; nevertheless I'll keep it open as a tracking issue (if a better, dedicated, tracking issue is created, this should be closed) but lock it.

**Comment by justinmk** [comment]
> Would be happy to look into it and submit a PR

vim/nvim calls the "cursor check" function(s) too often. Long ago I removed a couple calls in normal.c and saw a huge improvement with cursor movement performance. I would start there.

I also think we should support (maybe optionally, idk) deferring layout/viewport updates during cursor movement, just allow the screen to be "imperfect" (stale highlights) while user is moving the cursor, and update only once every ~300ms.
